File: lob/book.py
from collections import deque
from bisect import bisect_left
from typing import Dict, Deque, List, Optional

from .types import Order, Fill

class LimitOrderBook:
    def __init__(self):
        self.bids: Dict[int, Deque[int]] = {}
        self.asks: Dict[int, Deque[int]] = {}
        
        self.bid_prices: List[int] = []
        self.ask_prices: List[int] = []
        
        self.orders: Dict[int, Order] = {}
    
    #Helpers
    def best_bid(self) -> Optional[int]:
        return self.bid_prices[-1] if self.bid_prices else None
    
    def best_ask(self) -> Optional[int]:
        return self.ask_prices[0] if self.ask_prices else None
    
    def mid(self) -> Optional[float]:
        if self.best_bid() is None or self.best_ask() is None:
            return None
        return 0.5 * (self.best_bid() + self.best_ask())
    
    def spread(self) -> Optional[int]:
        if self.best_bid() is None or self.best_ask() is None:
            return None
        return self.best_ask() - self.best_bid()
    
    def depth(self, side: str, levels: int = 5) -> List[tuple[int, int]]:
        if levels <= 0:
            return []

        if side == "BID":
            book = self.bids
            prices = self.bid_prices
            # best bid is highest price
            pxs = list(reversed(prices[-levels:]))
        elif side == "ASK":
            book = self.asks
            prices = self.ask_prices
            # best ask is lowest price
            pxs = prices[:levels]
        else:
            raise ValueError(f"Invalid side, must be BID or ASK, got: {side}")

        out: List[tuple[int, int]] = []
        for px in pxs:
            q = book.get(px)
            if not q:
                continue
            total = 0
            for oid in q:
                o = self.orders.get(oid)
                if o is not None:
                    total += o.qty
            out.append((px, total))
        return out
    
    def imbalance(self, levels: int = 1) -> float:
        bid_vol = sum(qty for _, qty in self.depth("BID", levels=levels))
        ask_vol = sum(qty for _, qty in self.depth("ASK", levels=levels))
        denom = bid_vol + ask_vol
        if denom == 0:
            return 0.0
        return (bid_vol - ask_vol) / denom
        
    
    #Core API
    
    def add_limit(self, order: Order) -> List[Fill]:
        fills: List[Fill] = []
        
        #Crossing
        if order.side == "BID":
            while order.qty > 0 and self.best_ask() is not None and order.px >= self.best_ask():
                fills.extend(self._match(order, "ASK"))
        else:
            while order.qty > 0 and self.best_bid() is not None and order.px <= self.best_bid():
                fills.extend(self._match(order, "BID"))
    
        #Remainder Rests
        if order.qty > 0:
            self._add_to_book(order)
        
        return fills
    
    def add_market(self, side: str, qty: int, ts: int) -> List[Fill]:
        dummy = Order(
            id = -1,
            side = side,
            px = 0,
            qty = qty,
            ts = ts,
            order_type = "MARKET"
        )
        
        fills: List[Fill] = []
        opp = "ASK" if side == "BID" else "BID"
        
        while dummy.qty > 0 and ((self.best_ask() if side == "BID" else self.best_bid()) is not None):
            fills.extend(self._match(dummy, opp))
        
        return fills
    
    def cancel(self, order_id: int, ts:int) -> bool:
        if order_id not in self.orders:
            return False
        
        order = self.orders.pop(order_id)
        book = self.bids if order.side == "BID" else self.asks
        prices = self.bid_prices if order.side == "BID" else self.ask_prices
        
        q = book[order.px]
        q.remove(order_id)
        
        if not q:
            del book[order.px]
            prices.remove(order.px)
            
        return True
    
    #Private
    
    def _add_to_book(self, order: Order):
        book = self.bids if order.side == "BID" else self.asks
        prices = self.bid_prices if order.side == "BID" else self.ask_prices
        
        if order.px not in book:
            book[order.px] = deque()
            idx = bisect_left(prices, order.px)
            prices.insert(idx, order.px)
            
        book[order.px].append(order.id)
        self.orders[order.id] = order
    
    def _match(self, incoming: Order, resting_side: str) -> List[Fill]:
        fills: List[Fill] = []
        
        book = self.asks if resting_side == "ASK" else self.bids
        prices = self.ask_prices if resting_side == "ASK" else self.bid_prices
        
        if not prices:
            return fills
        
        px = prices[0] if resting_side == "ASK" else prices[-1]
        queue = book[px]
        
        while incoming.qty > 0 and queue:
            order_id = queue[0]
            resting = self.orders[order_id]
            
            traded = min(incoming.qty, resting.qty)
            
            incoming.qty -= traded
            resting.qty -= traded
            
            fills.append(
                Fill(
                    order_id = resting.id,
                    px = px,
                    qty = traded,
                    ts = incoming.ts,
                    liquidity = "MAKER",
                )
            )
            
            if resting.qty == 0:
                queue.popleft()
                del self.orders[order_id]
        if not queue:
            del book[px]
            prices.remove(px)
        
        return fills

File: cpp/src/book.cpp
#include "book.hpp"

#include <algorithm>
#include <cmath>

std::optional<int64_t> LimitOrderBook::best_bid() const {
    if (bids_.empty()) {
        return std::nullopt;
    }
    return bids_.begin()->first;
}

std::optional<int64_t> LimitOrderBook::best_ask() const {
    if (asks_.empty()) {
        return std::nullopt;
    }
    return asks_.begin()->first;
}

std::optional<double> LimitOrderBook::mid() const {
    if (!best_bid().has_value() || !best_ask().has_value()) {
        return std::nullopt;
    }
    return 0.5 * (static_cast<double>(*best_bid()) + static_cast<double>(*best_ask()));
}

std::optional<int64_t> LimitOrderBook::spread() const {
    if (!best_bid().has_value() || !best_ask().has_value()) {
        return std::nullopt;
    }
    return *best_ask() - *best_bid();
}

std::vector<std::pair<int64_t, int64_t>> LimitOrderBook::depth(Side side, int levels) const {
    if (levels <= 0) {
        return {};
    }

    std::vector<std::pair<int64_t, int64_t>> out;
    out.reserve(static_cast<size_t>(levels));

    if (side == Side::Bid) {
        int count = 0;
        for (const auto& [px, queue] : bids_) {
            int64_t total = 0;
            for (int64_t oid : queue) {
                auto it = orders_.find(oid);
                if (it != orders_.end()) {
                    total += it->second.qty;
                }
            }
            out.emplace_back(px, total);
            count += 1;
            if (count >= levels) {
                break;
            }
        }
    } else {
        int count = 0;
        for (const auto& [px, queue] : asks_) {
            int64_t total = 0;
            for (int64_t oid : queue) {
                auto it = orders_.find(oid);
                if (it != orders_.end()) {
                    total += it->second.qty;
                }
            }
            out.emplace_back(px, total);
            count += 1;
            if (count >= levels) {
                break;
            }
        }
    }

    return out;
}

double LimitOrderBook::imbalance(int levels) const {
    int64_t bid_vol = 0;
    int64_t ask_vol = 0;
    for (const auto& [_, qty] : depth(Side::Bid, levels)) {
        bid_vol += qty;
    }
    for (const auto& [_, qty] : depth(Side::Ask, levels)) {
        ask_vol += qty;
    }
    int64_t denom = bid_vol + ask_vol;
    if (denom == 0) {
        return 0.0;
    }
    return static_cast<double>(bid_vol - ask_vol) / static_cast<double>(denom);
}

std::vector<Fill> LimitOrderBook::add_limit(Order order) {
    std::vector<Fill> fills;

    if (order.side == Side::Bid) {
        while (order.qty > 0 && best_ask().has_value() && order.px >= *best_ask()) {
            auto f = match(order, Side::Ask);
            fills.insert(fills.end(), f.begin(), f.end());
        }
    } else {
        while (order.qty > 0 && best_bid().has_value() && order.px <= *best_bid()) {
            auto f = match(order, Side::Bid);
            fills.insert(fills.end(), f.begin(), f.end());
        }
    }

    if (order.qty > 0) {
        add_to_book(order);
    }

    return fills;
}

std::vector<Fill> LimitOrderBook::add_market(Side side, int64_t qty, int64_t ts) {
    Order dummy{
        -1,
        side,
        0,
        qty,
        ts,
        "SIM",
        OrderType::Market,
    };

    std::vector<Fill> fills;
    Side opp = (side == Side::Bid) ? Side::Ask : Side::Bid;

    while (dummy.qty > 0) {
        if (opp == Side::Ask && !best_ask().has_value()) {
            break;
        }
        if (opp == Side::Bid && !best_bid().has_value()) {
            break;
        }
        auto f = match(dummy, opp);
        fills.insert(fills.end(), f.begin(), f.end());
    }

    return fills;
}

bool LimitOrderBook::cancel(int64_t order_id, int64_t /*ts*/) {
    auto it = orders_.find(order_id);
    if (it == orders_.end()) {
        return false;
    }

    const Order order = it->second;

    if (order.side == Side::Bid) {
        auto book_it = bids_.find(order.px);
        if (book_it == bids_.end()) {
            return false;
        }
        auto& queue = book_it->second;
        auto qit = std::find(queue.begin(), queue.end(), order_id);
        if (qit == queue.end()) {
            return false;
        }
        queue.erase(qit);
        if (queue.empty()) {
            bids_.erase(book_it);
        }
    } else {
        auto book_it = asks_.find(order.px);
        if (book_it == asks_.end()) {
            return false;
        }
        auto& queue = book_it->second;
        auto qit = std::find(queue.begin(), queue.end(), order_id);
        if (qit == queue.end()) {
            return false;
        }
        queue.erase(qit);
        if (queue.empty()) {
            asks_.erase(book_it);
        }
    }

    // Only erase from orders_ after we successfully removed it from the book structure.
    orders_.erase(it);
    return true;
}

size_t LimitOrderBook::order_count() const {
    return orders_.size();
}

std::vector<int64_t> LimitOrderBook::all_order_ids() const {
    std::vector<int64_t> ids;
    ids.reserve(orders_.size());
    for (const auto& kv : orders_) {
        ids.push_back(kv.first);
    }
    return ids;
}

std::vector<int64_t> LimitOrderBook::order_ids_at_price(Side side, int64_t px) const {
    std::vector<int64_t> ids;
    if (side == Side::Bid) {
        auto it = bids_.find(px);
        if (it == bids_.end()) {
            return ids;
        }
        ids.assign(it->second.begin(), it->second.end());
    } else {
        auto it = asks_.find(px);
        if (it == asks_.end()) {
            return ids;
        }
        ids.assign(it->second.begin(), it->second.end());
    }
    return ids;
}

void LimitOrderBook::add_to_book(const Order& order) {
    if (order.side == Side::Bid) {
        auto& queue = bids_[order.px];
        queue.push_back(order.id);
    } else {
        auto& queue = asks_[order.px];
        queue.push_back(order.id);
    }
    orders_[order.id] = order;
}

std::vector<Fill> LimitOrderBook::match(Order& incoming, Side resting_side) {
    std::vector<Fill> fills;

    if (resting_side == Side::Ask) {
        if (asks_.empty()) {
            return fills;
        }
        auto it = asks_.begin();
        int64_t px = it->first;
        auto& queue = it->second;

        while (incoming.qty > 0 && !queue.empty()) {
            int64_t order_id = queue.front();
            auto rest_it = orders_.find(order_id);
            if (rest_it == orders_.end()) {
                queue.pop_front();
                continue;
            }
            Order& resting = rest_it->second;
            int64_t traded = std::min(incoming.qty, resting.qty);

            incoming.qty -= traded;
            resting.qty -= traded;

            fills.push_back(Fill{resting.id, px, traded, Liquidity::Maker, incoming.ts});

            if (resting.qty == 0) {
                queue.pop_front();
                orders_.erase(rest_it);
            }
        }

        if (queue.empty()) {
            asks_.erase(it);
        }
    } else {
        if (bids_.empty()) {
            return fills;
        }
        auto it = bids_.begin();
        int64_t px = it->first;
        auto& queue = it->second;

        while (incoming.qty > 0 && !queue.empty()) {
            int64_t order_id = queue.front();
            auto rest_it = orders_.find(order_id);
            if (rest_it == orders_.end()) {
                queue.pop_front();
                continue;
            }
            Order& resting = rest_it->second;
            int64_t traded = std::min(incoming.qty, resting.qty);

            incoming.qty -= traded;
            resting.qty -= traded;

            fills.push_back(Fill{resting.id, px, traded, Liquidity::Maker, incoming.ts});

            if (resting.qty == 0) {
                queue.pop_front();
                orders_.erase(rest_it);
            }
        }

        if (queue.empty()) {
            bids_.erase(it);
        }
    }

    return fills;
}
